File: solutions/974.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        unordered_map<int, int> mp;
        int res = 0;
        int sum = 0;
        mp[0]++;
        for (auto& num : nums) {
            sum += num;
            int mod = ((sum % k) + k) % k;
            res += mp[mod];
            mp[mod]++;
        }
        return res;
    }
};

File: rust_solutions/974.rs
impl Solution {
    pub fn subarrays_div_by_k(nums: Vec<i32>, k: i32) -> i32 {
        let mut mp = std::collections::HashMap::new();
        let mut res: i32 = 0;
        let mut sum: i32 = 0;

        mp.insert(0, 1);

        for num in nums {
            sum += num;
            let m = ((sum % k) + k) % k;
            if let Some(count) = mp.get(&m) {
                res += count;
            }
            *mp.entry(m).or_insert(0) += 1;
        }
        return res;
    }
}
